streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java

```diff
@@ -0,0 +1,293 @@
+        final KafkaStreams client2 = createClient(builder, streamsConfiguration());
+        final List<KafkaStreams> clients = Arrays.asList(client1, client2);
+
+        final AtomicLong restoreStartOffset = new AtomicLong(-1);
+        clients.forEach(client -> {
+            client.setGlobalStateRestoreListener(createTrackingRestoreListener(restoreStartOffset));
+            client.start();
+        });
+        waitForClientsToEnterRunningState(clients, 60, TimeUnit.SECONDS);
+
+        produceValueRange(key, 0, 10);
+
+        assertThat("all messages in the first batch were processed in a timely manner",
+            semaphore.tryAcquire(numMessages, 60, TimeUnit.SECONDS));
+
+        assertThat("no restore has occurred", restoreStartOffset.get(), is(equalTo(-1L)));
+    }
+
+    @Test
+    public void shouldApplyUpdatesToStandbyStore() throws InterruptedException, ExecutionException {
+        final int batch1NumMessages = 100;
+        final int batch2NumMessages = 100;
+        final int key = 1;
+        final Semaphore semaphore = new Semaphore(0);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+        builder
+            .table(INPUT_TOPIC_NAME, Consumed.with(Serdes.Integer(), Serdes.Integer()),
+                Materialized.<Integer, Integer, KeyValueStore<Bytes, byte[]>>as(TABLE_NAME)
+                    .withCachingDisabled())
+            .toStream()
+            .peek((k, v) -> semaphore.release());
+
+        final KafkaStreams client1 = createClient(builder, streamsConfiguration());
+        final KafkaStreams client2 = createClient(builder, streamsConfiguration());
+        final List<KafkaStreams> clients = Arrays.asList(client1, client2);
+
+        final AtomicLong restoreStartOffset = new AtomicLong(-1L);
+        clients.forEach(client -> {
+            client.setGlobalStateRestoreListener(createTrackingRestoreListener(restoreStartOffset));
+            client.start();
+        });
+        waitForClientsToEnterRunningState(clients, 60, TimeUnit.SECONDS);
+
+        produceValueRange(key, 0, batch1NumMessages);
+
+        assertThat("all messages in the first batch were processed in a timely manner",
+            semaphore.tryAcquire(batch1NumMessages, 60, TimeUnit.SECONDS));
+
+        final ReadOnlyKeyValueStore<Integer, Integer> store1 = client1
+            .store(TABLE_NAME, QueryableStoreTypes.keyValueStore());
+
+        final ReadOnlyKeyValueStore<Integer, Integer> store2 = client2
+            .store(TABLE_NAME, QueryableStoreTypes.keyValueStore());
+
+        final boolean client1WasFirstActive;
+        if (store1.get(key) != null) {
+            client1WasFirstActive = true;
+        } else {
+            assertThat("data from the job was sent to the store", store2.get(key) != null);
+            client1WasFirstActive = false;
+        }
+
+        assertThat("no restore has occurred", restoreStartOffset.get(), is(equalTo(-1L)));
```

Comment:
This seems to be covered by the first test already -- why do we want to check this condition in this test, too?

Reply:
The justification for this would be that the second time we do the query we're absolutely sure it wasn't tampered with by the first pass. I'd be inclined to pull the first test over removing this assertion. Let me know if you have a preference.

Reply:
I kept the original test and put a note about why we're checking it for this test.